cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java

```diff
@@ -1659,13 +1653,72 @@ private String generatePerBrokerConfiguration(NodeRef node, KafkaPool pool, Map<
                         .withAuthorization(cluster, authorization)
                         .withCruiseControl(cluster, ccMetricsReporter, node.broker())
                         .withUserConfiguration(configuration, node.broker() && ccMetricsReporter != null);
+        buildBrokerConfiguration(node, pool, builder);
+        return builder.build().trim();
+    }
 
-        if (useKRaft) {
-            builder.withKRaft(cluster, namespace, pool.processRoles, nodes());
-        } else {
-            builder.withZookeeper(cluster);
+    /**
+     * Adds ZooKeeper and/or KRaft connection and/or ZooKeeper migration depending on the role of the node (broker or controller)
+     * and the Kafka metadata configuration state.
+     * This method actually implements the following table.
+     *
+     * +----------------+--------------+--------------+
+     * |                | Controller   | Broker       |
+     * +----------------+--------------+--------------+ ---> strimzi.io/kraft = disabled
+     * | ZK             | -            | Zk           |
+     * +----------------+--------------+--------------+ ---> strimzi.io/kraft = migration
+     * | PRE_MIGRATION  | KRaft        | Zk           |
+     * |                | Zk           |              | ---> controllers deployed
+     * |                | Zk-migration |              |
+     * +----------------+--------------+--------------+
+     * | MIGRATION      | KRaft        | KRaft        |
+     * |                | Zk           | Zk           | ---> brokers rolled
+     * |                | Zk-migration | Zk-migration |
+     * +----------------+--------------+--------------+ ---> strimzi.io/kraft = enabled
+     * | POST_MIGRATION | KRaft        | KRaft        |
+     * |                | Zk           |              | ---> brokers rolled
+     * |                | Zk-migration |              |
+     * +----------------+--------------+--------------+
+     * | KRAFT          | KRaft        | KRaft        | ---> controllers rolled
+     * +----------------+--------------+--------------+
```

Comment:
Very nice description of the inner logic behind migration 👍 better than 1000 words!

Reply:
Taking into account that it's not the internal state machine. This is a way to map, depending on the migration state, which configuration are needed on controller and broker.